templates/federated_reporting/mrproper.py

```diff
@@ -0,0 +1,244 @@
+        response = feederApi.put(
+            "user",
+            "mrproper",
+            {
+                "description": "mrproper Federated Host Cleanup user",
+                "email": "{}".format(email),
+                "password": "{}".format(mrproper_password),
+                "roles": ["mrproper"],
+            },
+        )
+        if response["status"] != 201:
+            print("Problem creating mrproper user on {}. {}".format(feederHostname, response))
+            sys.exit(1)
+        write_secret(MRPROPER_SECRET_PATH, mrproper_password)
+
+
+def main():
+    if not os.path.exists(MRPROPER_SECRET_PATH):
+        if sys.stdout.isatty():
+            interactive_setup()
+        else:
+            print(
+                "{} requires manual setup, please run as root interactively.".format(
+                    os.path.basename(__file__)
+                )
+            )
+            sys.exit(1)
+
+    mrproper_password = read_secret(MRPROPER_SECRET_PATH)
+    api = NovaApi(
+        api_user="mrproper", api_password=mrproper_password
+    )  # defaults to localhost
+    response = api.fr_hub_status()
+    if not (
+        response["status"] == 200
+        and response["role"] == "superhub"
+        and response["configured"]
+    ):
+        print(
+            "{} can only be run on a Federated Reporting hub configured to be superhub".format(
+                os.path.basename(__file__)
+            )
+        )
+        sys.exit(1)
+
+    response = api.fr_remote_hubs()
+    for hub in response['hubs']:
+        if hub["role"] != "feeder":
+            continue
+
+        feeder_hostkey = hub["hostkey"]
+
+        sql = """
+SELECT DISTINCT hostkey
+FROM hosts
+WHERE EXISTS(
+  SELECT 1 FROM __hubs WHERE hostkey = '{}'
+  AND __hubs.hub_id = hosts.hub_id
+) AND EXISTS(
+  SELECT 1 FROM lastseenhosts ls
+  JOIN (
+    SELECT hostkey, max(lastseentimestamp) as newesttimestamp
+    FROM lastseenhosts
+    WHERE lastseendirection = 'INCOMING'
+    GROUP BY hostkey
+  ) as newest
+  ON ls.hostkey = newest.hostkey
+  AND ls.lastseentimestamp = newest.newesttimestamp
+  AND ls.hostkey = hosts.hostkey
+  AND ls.hub_id != hosts.hub_id
+)""".format(
+            feeder_hostkey
+        )
+
+        response = api.query(sql)
+        if response["status"] != 200:
+            print(
+                "Unable to query for deletion candidates. Response was {}".format(
+                    response
+                )
+            )
+            sys.exit(1)
+        print("Hosts to delete on {} are {}".format(hub["ui_name"], response["rows"]))
+
+        hosts_to_delete = response["rows"]
+        feederHostname = hub["ui_name"]
+        feederApi = NovaApi(
+            api_user="mrproper",
+            api_password=mrproper_password,
+            cert_path=CERT_PATH,
+            hostname=feederHostname,
+        )
+        for row in hosts_to_delete:
+            host_to_delete = row[0]
```

Comment:
I'd still like to see an explanation of why this `[0]` is needed.

Reply:
Odd. I explained it another place. Maybe in nova_api.py? The query results are kind of weird in a way, the reason is because the row is a list of column values and in this case we are only selecting one. So I can add a comment here about that. :+1: